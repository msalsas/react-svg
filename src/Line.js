var React = require('react');
var PropTypes = React.PropTypes;
var Component = React.Component;

class Line extends Component {
    static propTypes = {
        length: PropTypes.number,
        degrees: PropTypes.number,
        strokeWidth: PropTypes.number,
        strokeColor: PropTypes.string
    };
	static defaultProps = {
        lineLength: 150,
		degrees: 45,
		strokeWidth: 5,
		strokeColor: '#fff'
    };
    render() {
        var degreesInRadians = (this.props.degrees - 90) * Math.PI / 180.0;
        var offsetY = 0;
        if (this.props.degrees > 90 && this.props.degrees <= 180) {
            degreesInRadians = (this.props.degrees + 90) * Math.PI / 180.0;
            offsetY = Math.abs(this.props.lineLength * Math.sin(degreesInRadians));
        }
        var width = Math.abs(this.props.lineLength * Math.cos(degreesInRadians)) + this.props.strokeWidth * 2;
        var height = Math.abs(this.props.lineLength * Math.sin(degreesInRadians)) + this.props.strokeWidth * 2;
        return (
            <svg width={width} height={height} xmlns="http://www.w3.org/2000/svg">
                <g>
                    <path d={this.describeLine(this.props.lineLength, degreesInRadians, offsetY, this.props.strokeWidth)} className="msalsas-line"
                    fill="none" strokeWidth={this.props.strokeWidth} stroke={this.props.strokeColor}/>
                </g>
            </svg>
        );
    }

    describeLine = function(lineLength, degreesInRadians, offsetY, strokeWidth) {
        var start = {
            x: strokeWidth,
            y: strokeWidth + offsetY
        };

        var end = {
            x: Math.abs(lineLength * Math.cos(degreesInRadians)) + strokeWidth,
            y: Math.abs(lineLength * Math.sin(degreesInRadians)) - offsetY + strokeWidth
        };

        return [
            "M", start.x, start.y,
            "L", end.x, end.y
        ].join(" ");
    };
}

export default Line;
