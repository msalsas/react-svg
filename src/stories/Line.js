import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import Line from '../Line';

storiesOf('Line', module)
    .add('orange 100px length, 45ยบ', () => {
        const props = {
            lineLength: 100,
            degrees: 45,
            strokeWidth: 3,
            strokeColor: 'orange'
        };

        return renderLine(props);
    })
    .add('yellow 50px length, 120ยบ', () => {
        const props = {
            lineLength: 50,
            degrees: 120,
            strokeWidth: 3,
            strokeColor: 'yellow'
        };

        return renderLine(props);
    })
    .add('red 80px length, 290ยบ', () => {
        const props = {
            lineLength: 80,
            degrees: 290,
            strokeWidth: 3,
            strokeColor: 'red'
        };

        return renderLine(props);
    })
    .add('green 130px length, 350ยบ', () => {
        const props = {
            lineLength: 130,
            degrees: 350,
            strokeWidth: 3,
            strokeColor: 'green'
        };

        return renderLine(props);
    })
    .add('without options', () => {
        const props = {};

        return renderLine(props);
    });


function renderLine(props) {
    return (
        <div className="svg-wrapper" style={{ background: 'gray' }}>
            <Line {...props} />
        </div>
    );
}