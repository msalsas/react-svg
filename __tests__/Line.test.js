import React from 'react';
import Line from '../src/Line';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('Test Line component', () => {
    const props = {
        lineLength   : 50,
        degrees      : 120,
        strokeWidth  : 10,
        strokeColor  : 'yellow',
        strokeLinecap: 'round'
    };
    const line = shallow((<Line {...props}/>));

    it('should be defined', () => {
        expect(line).toBeDefined();
    });
    it('should render correctly', () => {
        expect(line).toMatchSnapshot();
    });
});
