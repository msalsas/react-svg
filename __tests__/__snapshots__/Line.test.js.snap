// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Line component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Line
    degrees={120}
    lineLength={50}
    strokeColor="yellow"
    strokeLinecap="round"
    strokeWidth={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <g>
        <path
          className="msalsas-line"
          d="M 10 35.00000000000001 L 53.30127018922193 10"
          fill="none"
          stroke="yellow"
          strokeLinecap="round"
          strokeWidth={10}
        />
      </g>,
      "height": 45.00000000000001,
      "width": 63.30127018922193,
      "xmlns": "http://www.w3.org/2000/svg",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <path
          className="msalsas-line"
          d="M 10 35.00000000000001 L 53.30127018922193 10"
          fill="none"
          stroke="yellow"
          strokeLinecap="round"
          strokeWidth={10}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "msalsas-line",
          "d": "M 10 35.00000000000001 L 53.30127018922193 10",
          "fill": "none",
          "stroke": "yellow",
          "strokeLinecap": "round",
          "strokeWidth": 10,
        },
        "ref": null,
        "rendered": null,
        "type": "path",
      },
      "type": "g",
    },
    "type": "svg",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <g>
          <path
            className="msalsas-line"
            d="M 10 35.00000000000001 L 53.30127018922193 10"
            fill="none"
            stroke="yellow"
            strokeLinecap="round"
            strokeWidth={10}
          />
        </g>,
        "height": 45.00000000000001,
        "width": 63.30127018922193,
        "xmlns": "http://www.w3.org/2000/svg",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <path
            className="msalsas-line"
            d="M 10 35.00000000000001 L 53.30127018922193 10"
            fill="none"
            stroke="yellow"
            strokeLinecap="round"
            strokeWidth={10}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "msalsas-line",
            "d": "M 10 35.00000000000001 L 53.30127018922193 10",
            "fill": "none",
            "stroke": "yellow",
            "strokeLinecap": "round",
            "strokeWidth": 10,
          },
          "ref": null,
          "rendered": null,
          "type": "path",
        },
        "type": "g",
      },
      "type": "svg",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
